addons/binding/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/protocol/RemoteControllerWebSocket.java

```diff
@@ -0,0 +1,678 @@
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        logger.debug("RemoteControllerWebSocket close");
+        closeConnection();
+    }
+
+    /**
+     * Websocket base class
+     *
+     * @author Arjan Mels
+     *
+     */
+    class WebSocketBase extends WebSocketAdapter {
+        boolean isConnecting = false;
+
+        @Override
+        public void onWebSocketClose(int statusCode, String reason) {
+            logger.debug("{} connection closed: {} - {}", this.getClass().getSimpleName(), statusCode, reason);
+            super.onWebSocketClose(statusCode, reason);
+            isConnecting = false;
+        }
+
+        @Override
+        public void onWebSocketError(Throwable error) {
+            logger.error("{} connection error: {}", this.getClass().getSimpleName(), error.getMessage());
+            super.onWebSocketError(error);
+            isConnecting = false;
+        }
+
+        void connect(URI uri) throws RemoteControllerException {
+            if (isConnecting || isConnected()) {
+                logger.trace("{} already connecting or connected", this.getClass().getSimpleName());
+                return;
+            }
+
+            try {
+                logger.debug("{} connecting to: {}", this.getClass().getSimpleName(), uri);
+
+                isConnecting = true;
+                client.connect(this, uri, new ClientUpgradeRequest());
+            } catch (Exception e) {
+                throw new RemoteControllerException(e);
+            }
+        }
+
+        @Override
+        public void onWebSocketConnect(Session session) {
+            logger.debug("{} connection established: {}", this.getClass().getSimpleName(),
+                    session.getRemoteAddress().getHostString());
+            super.onWebSocketConnect(session);
+
+            isConnecting = false;
+        }
+
+        void close() {
+            getSession().close();
+        }
+
+        void sendCommand(String cmd) {
+            try {
+                // retry openening connection just in case
+                openConnection();
+
+                if (isConnected()) {
+                    getRemote().sendString(cmd);
+                    logger.trace("{}: sendCommand: {}", this.getClass().getSimpleName(), cmd);
+                } else {
+                    logger.warn("{} sending command while socket not connected: {}", this.getClass().getSimpleName(),
+                            cmd);
+                }
+            } catch (Exception e) {
+                logger.error("{}: cannot send command", this.getClass().getSimpleName(), e);
+            }
+        }
+
+        @Override
+        public void onWebSocketText(String str) {
+            logger.trace("{}: onWebSocketText: {}", this.getClass().getSimpleName(), str);
+        }
+
+    }
+
+    // temporary storage for source app. Will be used as value for the sourceApp channel when information is complete
+    private String currentSourceApp = null;

```

Comment:
Please keep your fields at the top or split up your class, this is not a place where I would expect fields.

Please also make the next one private.

Reply:
Done